File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/ChatCommand.cs
using Atlasd.Battlenet.Protocols.Game.ChatCommands;
using Atlasd.Daemon;
using Atlasd.Localization;
using System;
using System.Collections.Generic;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game
{
    class ChatCommand
    {
        public List<string> Arguments { get; protected set; }
        public byte[] RawBuffer { get; protected set; }

        /**
         * <param name="rawBuffer">The byte array from the client, stripped of the slash and null-terminator. Value may be used during invocation of a ChatCommand.</param>
         */
        public ChatCommand(byte[] rawBuffer, List<string> arguments)
        {
            Arguments = arguments;
            RawBuffer = rawBuffer;
        }

        public virtual bool CanInvoke(ChatCommandContext context)
        {
            return false;
        }

        public virtual void Invoke(ChatCommandContext context)
        {
            throw new NotSupportedException("Base ChatCommand class does not Invoke()");
        }

        public static ChatCommand FromByteArray(byte[] text)
        {
            return Parse(Encoding.UTF8.GetString(text), text);
        }

        public static ChatCommand FromString(string text)
        {
            return Parse(text, Encoding.UTF8.GetBytes(text));
        }

        private static ChatCommand Parse(string text, byte[] raw)
        {
            var args = new List<string>(text.Split(' '));

            var cmd = args[0];
            args.RemoveAt(0);

            // Calculates and removes (cmd+' ') from (raw) which prints into (_raw):
            var stripSize = cmd.Length + (text.Length - cmd.Length > 0 ? 1 : 0);
            var _raw = raw[stripSize..];

            switch (cmd.ToLowerInvariant())
            {
                case "admin":
                    return new AdminCommand(_raw, args);
                case "away":
                    return new AwayCommand(_raw, args);
                case "ban":
                    return new BanCommand(_raw, args);
                case "clan":
                    return new ClanCommand(_raw, args);
                case "channel":
                case "join":
                case "j":
                    return new JoinCommand(_raw, args);
                case "designate":
                    return new DesignateCommand(_raw, args);
                case "emote":
                case "me":
                    return new EmoteCommand(_raw, args);
                case "friends":
                case "friend":
                case "f":
                    return new FriendCommand(_raw, args);
                case "help":
                case "?":
                    return new HelpCommand(_raw, args);
                case "ignore":
                case "squelch":
                    return new SquelchCommand(_raw, args);
                case "kick":
                    return new KickCommand(_raw, args);
                case "rejoin":
                case "rj":
                    return new ReJoinCommand(_raw, args);
                case "time":
                    return new TimeCommand(_raw, args);
                case "unban":
                    return new UnBanCommand(_raw, args);
                case "unignore":
                case "unsquelch":
                    return new UnsquelchCommand(_raw, args);
                case "users":
                    return new UsersCommand(_raw, args);
                case "version":
                case "ver":
                    return new VersionCommand(_raw, args);
                case "whereis":
                case "where":
                case "whois":
                    return new WhereIsCommand(_raw, args);
                case "whisper":
                case "msg":
                case "m":
                case "w":
                    return new WhisperCommand(_raw, args);
                case "who":
                    return new WhoCommand(_raw, args);
                case "whoami":
                    return new WhoAmICommand(_raw, args);
                default:
                    return new InvalidCommand(_raw, args);
            }
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/ChatCommand.vb
Imports AtlasV.Battlenet.Protocols.Game.ChatCommands
Imports AtlasV.Daemon
Imports AtlasV.Localization
Imports System
Imports System.Collections.Generic
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game
    Class ChatCommand
        Public Property Arguments As List(Of String)
        Public Property RawBuffer As Byte()

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            Arguments = varArguments
            RawBuffer = varRawBuffer
        End Sub

        Public Overridable Function CanInvoke(ByVal context As ChatCommandContext) As Boolean
            Return False
        End Function

        Public Overridable Sub Invoke(ByVal context As ChatCommandContext)
            Throw New NotSupportedException("Base ChatCommand class does not Invoke()")
        End Sub

        Public Shared Function FromByteArray(ByVal text As Byte()) As ChatCommand
            Return Parse(Encoding.UTF8.GetString(text), text)
        End Function

        Public Shared Function FromString(ByVal text As String) As ChatCommand
            Return Parse(text, Encoding.UTF8.GetBytes(text))
        End Function

        Private Shared Function Parse(ByVal text As String, ByVal raw As Byte()) As ChatCommand
            Dim args = New List(Of String)(text.Split(" "c))
            Dim cmd = args(0)
            args.RemoveAt(0)
            Dim stripSize = cmd.Length + (If(text.Length - cmd.Length > 0, 1, 0))
            Dim _raw = raw.Skip(stripSize).ToArray()

            Select Case cmd
                Case "admin"
                    Return New AdminCommand(_raw, args)
                Case "away"
                    Return New AwayCommand(_raw, args)
                Case "ban"
                    Return New BanCommand(_raw, args)
                Case "clan"
                    Return New ClanCommand(_raw, args)
                Case "channel", "join", "j"
                    Return New JoinCommand(_raw, args)
                Case "designate"
                    Return New DesignateCommand(_raw, args)
                Case "emote", "me"
                    Return New EmoteCommand(_raw, args)
                Case "friends", "friend", "f"
                    Return New FriendCommand(_raw, args)
                Case "help", "?"
                    Return New HelpCommand(_raw, args)
                Case "ignore", "squelch"
                    Return New SquelchCommand(_raw, args)
                Case "kick"
                    Return New KickCommand(_raw, args)
                Case "rejoin", "rj"
                    Return New ReJoinCommand(_raw, args)
                Case "time"
                    Return New TimeCommand(_raw, args)
                Case "unban"
                    Return New UnBanCommand(_raw, args)
                Case "unignore", "unsquelch"
                    Return New UnsquelchCommand(_raw, args)
                Case "users"
                    Return New UsersCommand(_raw, args)
                Case "version", "ver"
                    Return New VersionCommand(_raw, args)
                Case "whereis", "where", "whois"
                    Return New WhereIsCommand(_raw, args)
                Case "whisper", "msg", "m", "w"
                    Return New WhisperCommand(_raw, args)
                Case "who"
                    Return New WhoCommand(_raw, args)
                Case "whoami"
                    Return New WhoAmICommand(_raw, args)
                Case Else
                    Return New InvalidCommand(_raw, args)
            End Select
        End Function
    End Class
End Namespace
